PyTrilinos: Revive PyTrilinos.Tpetra

@trilinos/pytrilinos 
@trilinos/tpetra 

## Expectations
When PyTrilinos and Tpetra are enabled, working Python wrappers for Tpetra should be generated

## Current Behavior
Recent changes to the Tpetra package have broken the `PyTrilinos.Tpetra` package

## Definition of Done
The Tpetra wrappers compile and the PyTrilinos test suite passes

## Possible Solution
Modifications to `packages/PyTrilinos/src/Tpetra.i` will need to be made, possibly other files, such as `PyTrilinos_Tpetra_Util.cpp`, to reflect the current state of the Tpetra package.

@mhoemmen Second question: when I try to compile my Tpetra wrappers, I get an error due to a `private` type:

    /Development/Trilinos/PYTPETRA/packages/PyTrilinos/src/TpetraPYTHON_wrap.cpp:8569:83: error: 'global_indices_array_type' is a private member of 'Tpetra::Classes::Map<int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >'

What is happening is that the `Map` method `getMyGlobalIndices` has a return type that is `private`. What are users expected to do? Use `auto`? I don't think this is an option for SWIG.

If that is the case, I can use a `%extend` SWIG directive to redefine the wrapper. I'm just curious if there its something easier to do.

--- a/packages/PyTrilinos/src/Tpetra.i
+++ b/packages/PyTrilinos/src/Tpetra.i
@@ -81,6 +81,9 @@ using Teuchos::ArrayRCP;
 
 // Tpetra include files
 #include "PyTrilinos_Tpetra_Headers.hpp"
+using Tpetra::Map;
+using Tpetra::Export;
+using Tpetra::Import;
 
 #ifdef HAVE_DOMI
 // Domi include files
@@ -639,7 +642,7 @@ __version__ = version()
 // Tpetra Map support //
 ////////////////////////
 
-%extend Tpetra::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >
+%extend Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >
 {
   Map()
   {
@@ -788,11 +791,11 @@ __version__ = version()
   }
 }
 
-%ignore Tpetra::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::Map;
-%ignore Tpetra::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getLocalElement;
-%ignore Tpetra::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getGlobalElement;
-%ignore Tpetra::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getRemoteIndexList;
-%ignore Tpetra::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getMyGlobalIndices;
+%ignore Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::Map;
+%ignore Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getLocalElement;
+%ignore Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getGlobalElement;
+%ignore Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getRemoteIndexList;
+%ignore Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD, PYTRILINOS_GLOBAL_ORD, DefaultNodeType >::getMyGlobalIndices;
 
 %include "Tpetra_Map_decl.hpp"
 
@@ -800,21 +803,21 @@ __version__ = version()
 // directives below are redundant, because it is the same as the
 // default template argument.  But SWIG is much more acurate when
 // comparing types when all template arguments are specified.
-%teuchos_rcp(Tpetra::Map< PYTRILINOS_LOCAL_ORD,
-                          PYTRILINOS_GLOBAL_ORD,
-                          DefaultNodeType >)
-%template(Map_default) Tpetra::Map< PYTRILINOS_LOCAL_ORD,
-                                    PYTRILINOS_GLOBAL_ORD,
-                                    DefaultNodeType >;
+%teuchos_rcp(Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD,
+                                   PYTRILINOS_GLOBAL_ORD,
+                                   DefaultNodeType >)
+%template(Map_default) Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD,
+                                             PYTRILINOS_GLOBAL_ORD,
+                                             DefaultNodeType >;
 %pythoncode
 {
 Map = Map_default
 }
 %inline
 %{
-  typedef Tpetra::Map< PYTRILINOS_LOCAL_ORD,
-                       PYTRILINOS_GLOBAL_ORD,
-                       DefaultNodeType > DefaultMapType;
+  typedef Tpetra::Classes::Map< PYTRILINOS_LOCAL_ORD,
+                                PYTRILINOS_GLOBAL_ORD,
+                                DefaultNodeType > DefaultMapType;
 %}
 
 /////////////////////////////
@@ -866,12 +869,12 @@ public:
 // Tpetra Export support //
 ///////////////////////////
 %include "Tpetra_Export_decl.hpp"
-%teuchos_rcp(Tpetra::Export< PYTRILINOS_LOCAL_ORD,
-                             PYTRILINOS_GLOBAL_ORD,
-                             DefaultNodeType >)
-%template(Export_default) Tpetra::Export< PYTRILINOS_LOCAL_ORD,
-                                          PYTRILINOS_GLOBAL_ORD,
-                                          DefaultNodeType >;
+%teuchos_rcp(Tpetra::Classes::Export< PYTRILINOS_LOCAL_ORD,
+                                      PYTRILINOS_GLOBAL_ORD,
+                                      DefaultNodeType >)
+%template(Export_default) Tpetra::Classes::Export< PYTRILINOS_LOCAL_ORD,
+                                                   PYTRILINOS_GLOBAL_ORD,
+                                                   DefaultNodeType >;
 %pythoncode
 {
 Export = Export_default
@@ -881,12 +884,12 @@ Export = Export_default
 // Tpetra Import support //
 ///////////////////////////
 %include "Tpetra_Import_decl.hpp"
-%teuchos_rcp(Tpetra::Import< PYTRILINOS_LOCAL_ORD,
-                             PYTRILINOS_GLOBAL_ORD,
-                             DefaultNodeType >)
-%template(Import_default) Tpetra::Import< PYTRILINOS_LOCAL_ORD,
-                                          PYTRILINOS_GLOBAL_ORD,
-                                          DefaultNodeType >;
+%teuchos_rcp(Tpetra::Classes::Import< PYTRILINOS_LOCAL_ORD,
+                                      PYTRILINOS_GLOBAL_ORD,
+                                      DefaultNodeType >)
+%template(Import_default) Tpetra::Classes::Import< PYTRILINOS_LOCAL_ORD,
+                                                   PYTRILINOS_GLOBAL_ORD,
+                                                   DefaultNodeType >;
 %pythoncode
 {
 Import = Import_default

--- a/packages/PyTrilinos/src/gen_teuchos_rcp.py.in
+++ b/packages/PyTrilinos/src/gen_teuchos_rcp.py.in
@@ -61,7 +61,8 @@ def get_input_files():
     os.chdir("${CMAKE_CURRENT_SOURCE_DIR}")
     result = glob.glob("Teuchos*.i")
     if (WITH_EPETRA):
-        result.extend(glob.glob("Epetra*.i"))
+        result.append("Epetra.i")
+        result.extend(glob.glob("Epetra_*.i"))
     if (WITH_TRIUTILS):
         result.extend(glob.glob("TriUtils*.i"))
     if (WITH_TPETRA):

--- a/packages/PyTrilinos/test/testTeuchos.py.in
+++ b/packages/PyTrilinos/test/testTeuchos.py.in
@@ -52,15 +52,15 @@ import unittest
 
 #
 # Parse the command-line arguments
-#from optparse import *
-#parser = OptionParser()
-#parser.add_option("-t", "--testharness", action="store_true",
-#                  dest="testharness", default=False,
-#                  help="test local build modules; prevent loading system-installed modules")
-#parser.add_option("-v", "--verbosity", type="int", dest="verbosity", default=2,
-#                  help="set the verbosity level [default 2]")
-#options,args = parser.parse_args()
-options = type("Options", (object,), {"testharness":True, "verbosity":2})
+from optparse import *
+parser = OptionParser()
+parser.add_option("-t", "--testharness", action="store_true",
+                  dest="testharness", default=False,
+                  help="test local build modules; prevent loading system-installed modules")
+parser.add_option("-v", "--verbosity", type="int", dest="verbosity", default=2,
+                  help="set the verbosity level [default 2]")
+options,args = parser.parse_args()
+
 #
 # Under normal usage, simply use 'from PyTrilinos import Teuchos'.  For testing,
 # we want to be able to control whether we import from the build directory or

